New aws_region_facts module to be able to get facts from AWS regions

Sodki opened the pull request:
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

aws_region_facts

##### ANSIBLE VERSION

```
ansible 2.3.0 (devel 28e39ca42f) last updated 2017/01/25 16:44:10 (GMT +100)
  config file = 
  configured module search path = Default w/o overrides
```

##### SUMMARY

This is a new module that allows to query EC2 and obtain a a list of regions. It's useful if you want to iterate through the regions in search of something specific. I wrote this module because it was something I needed to do.

This implementation uses the same skeleton as `ec2_group_facts`.

Commit:
New aws_region_facts module to be able to get facts from AWS regions

j-carl commented on lib/ansible/modules/cloud/amazon/ec2_region_facts.py:
```diff
@@ -0,0 +1,125 @@
+
+
+from ansible.module_utils.basic import *
+from ansible.module_utils.ec2 import *
```
Please change module imports to the new style. Take a look at [vmware_guest.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_guest.py) for an example.

willthames commented on lib/ansible/modules/cloud/amazon/ec2_region_facts.py:
```diff
@@ -0,0 +1,125 @@
+
+RETURN = '''
+regions:
+    description: Regions that match the provided filters. Each element consists of a dict with all the information related to that region.
```
Please wrap lines at somewhere around 120 chars. You might want to use

```
  description: >
     Regions that match...
```

willthames commented on lib/ansible/modules/cloud/amazon/ec2_region_facts.py:
```diff
@@ -0,0 +1,125 @@
+        module.fail_json(msg=e.message, exception=traceback.format_exc(e))
+
+    # Turn the boto3 result into ansible_friendly_snaked_names
+    snaked_regions = []
```
this section can be simplified to

```
module.exit_json(regions=[camel_dict_to_snake_dict(region) for region in regions['Regions']])
```

willthames commented on lib/ansible/modules/cloud/amazon/ec2_region_facts.py:
```diff
@@ -0,0 +1,125 @@
+'''
+
+
+try:
```
If you import `HAS_BOTO3` from ansible.module_utils.ec2 this section can be

```
try: 
    from botocore.exceptions import ClientError
except ImportError:
    pass # will be detected by imported HAS_BOTO3
```

willthames commented:
I expect all new AWS modules to pass flake8, and most of my comments will help with that.

Sodki commented:
@willthames I've renamed the module and implemented your change requests. flake8 is happy on this one as well.

Sodki commented:
ready_for_review

willthames commented on lib/ansible/modules/cloud/amazon/aws_region_facts.py:
```diff
@@ -0,0 +1,120 @@
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import get_aws_connection_info, ec2_argument_spec, boto3_conn, \
+    ansible_dict_to_boto3_filter_list, camel_dict_to_snake_dict, HAS_BOTO3
```
I prefer another from line to a trailing slash.

So line 69 would be
```
from ansible.module_utils.ec2 import ansible_dict_to_boto3_filter_list, camel_dict_to_snake_dict, HAS_BOTO3
```

willthames commented on lib/ansible/modules/cloud/amazon/aws_region_facts.py:
```diff
@@ -0,0 +1,120 @@
+        module.fail_json(msg="region must be specified")
+
+    # Replace filter key underscores with dashes, for compatibility
+    sanitized_filters = module.params.get("filters")
```
```
sanitized_filters = [key.replace("_", "-") for key in module.params.get('filters')]
```

willthames commented:
shipit

I've made some general comments but they're subjective style rather than essential changes. If you do choose to make those changes, I'll reapply this shipit.

Sodki commented:
@willthames I've applies the changes you mentioned, makes the code look nicer.

ready_for_review

mattclay commented on lib/ansible/modules/cloud/amazon/aws_region_facts.py:
```diff
@@ -0,0 +1,118 @@
+        module.fail_json(msg='region must be specified')
+
+    # Replace filter key underscores with dashes, for compatibility
+    sanitized_filters = {k.replace('_', '-'):v for k,v in module.params.get('filters').items()}
```
CI failure due to use of dict comprehension, which isn't supported on python 2.6:

```
2017-02-17 15:38:36 Compiling ./lib/ansible/modules/cloud/amazon/aws_region_facts.py ...
2017-02-17 15:38:36 SyntaxError: ('invalid syntax', ('./lib/ansible/modules/cloud/amazon/aws_region_facts.py', 105, 50, "    sanitized_filters = {k.replace('_', '-'):v for k,v in module.params.get('filters').items()}\n")
```

See here for an alternative: http://stackoverflow.com/questions/21069668/alternative-to-dict-comprehension-prior-to-python-2-7

Sodki replied:
Thank you so much, I've fixed it now.

Sodki commented:
ready_for_review

willthames commented:
shipit

Sodki commented:
ready_for_review

s-hertel committed:
Fix copyright

Handle BotoCoreError too

Remove exception handling around around boto3_conn

Fix docs

Fix version_added

s-hertel committed:
Update metadata_version

Fix return doc

mtpereira commented:
This looks helpful, ship it! :)
